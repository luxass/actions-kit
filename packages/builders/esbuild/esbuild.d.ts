declare module "actions-kit/config" {
	import type { BuildOptions as ESBuildBuildOptions } from "esbuild";

	interface ActionsKitConfig {
		esbuild?: ESBuildBuildOptions;
	}
}

export {};
