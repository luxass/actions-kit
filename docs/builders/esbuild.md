# ESBuild Builder

Build GitHub Actions using ESBuild.

## Installation

::: code-group
  ```bash [pnpm]
  pnpm add -D actions-kit @actions-sdk/esbuild-builder
  ```

  ```bash [yarn]
  yarn add -D actions-kit @actions-sdk/esbuild-builder
  ```

  ```bash [npm]
  npm install -D actions-kit @actions-sdk/esbuild-builder
  ```

  ```bash [bun]
  bun install -D actions-kit @actions-sdk/esbuild-builder
  ```
:::
