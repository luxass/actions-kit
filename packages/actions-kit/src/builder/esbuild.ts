import { inferModuleType, inferOutputFilename } from "../utils";
import type { Config } from "../config";
import actionsKit from "unplugin-actions-kit/esbuild";
import { join } from "node:path";

export async function build(config: Config) {
	const esbuild = await import("esbuild").then((m) => m.build);
	const outputFileName = await inferOutputFilename(config);
	const libraryType = await inferModuleType(config, outputFileName);

	const esbuildConfig = config.esbuild;

	const a = await esbuild({
		bundle: true,
		platform: "node",
		target: "node20",
		format: libraryType,
		outfile: join(process.cwd(), "dist", outputFileName),
		plugins: [
			actionsKit({
				// TODO: allow users to specify it.
				actionPath: join(process.cwd(), "./action.yml"),
				inject: true,
			}),
		],
	});
}
