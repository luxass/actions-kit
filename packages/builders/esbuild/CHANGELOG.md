# @actions-sdk/esbuild-builder

## 0.2.3

### Patch Changes

- Updated dependencies [[`ae1a6ec`](https://github.com/luxass/actions-kit/commit/ae1a6ec04ad71f0adda06439877d65182719ba9a)]:
  - unplugin-actions-kit@0.2.3
  - actions-kit@0.2.4

## 0.2.2

### Patch Changes

- [`efedd0c`](https://github.com/luxass/actions-kit/commit/efedd0cf1448c1d480cddde2ef43a3939b325be6) Thanks [@luxass](https://github.com/luxass)! - run build before publish

- Updated dependencies [[`efedd0c`](https://github.com/luxass/actions-kit/commit/efedd0cf1448c1d480cddde2ef43a3939b325be6)]:
  - actions-kit@0.2.3
  - unplugin-actions-kit@0.2.2

## 0.2.1

### Patch Changes

- [#23](https://github.com/luxass/actions-kit/pull/23) [`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b) Thanks [@luxass](https://github.com/luxass)! - fix package exports and add missing cli entrypoint for `actions-kit`

- Updated dependencies [[`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b)]:
  - actions-kit@0.2.2

## 0.2.0

### Minor Changes

- [`bcf5300`](https://github.com/luxass/actions-kit/commit/bcf53008a191840aee634e040f4d9c53cfbf2a8b) Thanks [@luxass](https://github.com/luxass)! - change how builders work, for more information https://actions-kit.pages.dev/builders/
