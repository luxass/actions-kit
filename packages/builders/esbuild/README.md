# @actions-kit/esbuild-builder

ESBuild builder for Actions Kit.
