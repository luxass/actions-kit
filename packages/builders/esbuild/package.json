{
  "name": "@actions-sdk/esbuild-builder",
  "version": "0.2.7",
  "description": "ESBuild builder for Actions Kit",
  "type": "module",
  "author": {
    "name": "Lucas Nørgård",
    "email": "lucasnrgaard@gmail.com",
    "url": "https://luxass.dev"
  },
  "packageManager": "pnpm@10.14.0",
  "license": "MIT",
  "homepage": "https://github.com/luxass/actions-kit",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/luxass/actions-kit.git",
    "directory": "packages/builders/esbuild"
  },
  "bugs": {
    "url": "https://github.com/luxass/actions-kit/issues"
  },
  "keywords": [
    "esbuild",
    "actions-kit",
    "github actions",
    "actions"
  ],
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.cts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsdown --tsconfig=./tsconfig.build.json",
    "dev": "tsdown --watch",
    "clean": "git clean -xdf dist node_modules",
    "lint": "eslint .",
    "typecheck": "tsc --noEmit"
  },
  "peerDependencies": {
    "actions-kit": "workspace:*",
    "esbuild": "catalog:"
  },
  "dependencies": {
    "defu": "catalog:",
    "unplugin-actions-kit": "workspace:*"
  },
  "devDependencies": {
    "@actions-kit/tsconfig": "workspace:*",
    "@actions-kit/tsdown-config": "workspace:*",
    "@luxass/eslint-config": "catalog:linting",
    "@types/node": "catalog:types",
    "actions-kit": "workspace:*",
    "esbuild": "catalog:",
    "eslint": "catalog:linting",
    "publint": "catalog:dev",
    "tsdown": "catalog:dev",
    "typescript": "catalog:dev",
    "vitest": "catalog:testing"
  }
}
