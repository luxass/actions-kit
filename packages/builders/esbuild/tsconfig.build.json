{
  "extends": ["../../../tsconfig.base.build.json"],
  "include": [
    "./esbuild.d.ts",
    "src/**/*.ts"
  ]
}
