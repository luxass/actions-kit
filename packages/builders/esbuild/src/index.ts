import { defineBuilder, type BuildOutput } from "actions-kit/builder";
import { inferModuleType, inferOutput } from "actions-kit/builder-utils";
import { defu } from "defu";
import type { BuildOptions as ESBuildBuildOptions } from "esbuild";
import { build } from "esbuild";
import { join } from "node:path";
import ESBuildActionsKit from "unplugin-actions-kit/esbuild";
import { getESBuildEntryPoint } from "./utils";

export default function esbuildBuilder(options: ESBuildBuildOptions = {}) {
	return defineBuilder({
		name: "esbuild",
		build: async ({ cwd, config }) => {
			const defaultEntryPoint = config.build?.input || join(cwd, "src/index.ts");
			const entryPoints = getESBuildEntryPoint(options.entryPoints, defaultEntryPoint);

			if (options.entryPoints) {
				options.entryPoints = undefined;
			}

			const { filename, dir } = await inferOutput(config);
			const libraryType = await inferModuleType(config, filename);

			const esbuildOptions = defu(
				{
					...options,
					// metafile should always be true
					metafile: true,
				},
				{
					entryPoints: [entryPoints],
					platform: "node",
					target: "node20",
					format: libraryType,
					bundle: true,
					outfile: join(dir, filename),
					metafile: true,
					plugins: [
						ESBuildActionsKit({
							actionPath: join(cwd, "./action.yml"),
							inject: config.inject,
							autocomplete: config.autocomplete,
						}),
					],
				} satisfies ESBuildBuildOptions,
			);

			const result = await build(
				// TODO: fix later
				esbuildOptions as ESBuildBuildOptions,
			);

			const metafile = result.metafile;

			if (!metafile) {
				throw new Error("metafile didn't generate");
			}

			const outputFiles = Object.entries(metafile.outputs);

			const output: BuildOutput[] = [];
			for (const [outputFile, outputMeta] of outputFiles) {
				output.push({
					name: outputFile,
					path: join(cwd, outputFile),
					size: outputMeta.bytes,
				});
			}

			return output;
		},
	});
}
