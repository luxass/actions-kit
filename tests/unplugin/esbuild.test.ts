import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { build } from "esbuild";
import { expect, it } from "vitest";
import { fromFileSystem, testdir } from "vitest-testdirs";
import ActionKitPlugin from "../../src/unplugin/esbuild";

it("expect `actions-kit.d.ts` to be generated", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const result = await build({
    entryPoints: [
      join(testdirPath, "index.ts"),
    ],
    platform: "node",
    format: "cjs",
    write: false,
    bundle: false,
    minifySyntax: false,
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
      }),
    ],
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(result).toBeDefined();
  expect(result.outputFiles).toBeDefined();
  const output = result.outputFiles[0];
  expect(output).toBeDefined();
  expect(output?.text).toBeDefined();
  expect(dtsOutput).toMatchSnapshot();
  expect(output?.text).toMatchSnapshot();
});

it("expect `actions-kit.d.ts` to include `ACTION_INPUTS`", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const result = await build({
    entryPoints: [
      join(testdirPath, "index.ts"),
    ],
    platform: "node",
    format: "cjs",
    write: false,
    bundle: false,
    minifySyntax: false,
    treeShaking: false,
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
        injectInputs: true,
      }),
    ],
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(result).toBeDefined();
  expect(result.outputFiles).toBeDefined();
  const output = result.outputFiles[0];
  expect(output).toBeDefined();
  expect(output?.text).toBeDefined();
  expect(dtsOutput).toMatchSnapshot();
  expect(output?.text).toMatchSnapshot();

  expect(output?.text).toContain("ACTION_INPUTS");
  expect(dtsOutput).toContain("ACTION_INPUTS");
});
