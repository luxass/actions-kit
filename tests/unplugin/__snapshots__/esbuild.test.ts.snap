// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect \`actions-kit.d.ts\` to be generated 1`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to be generated 2`] = `
""use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var core = __toESM(require("@actions/core"), 1);
async function run() {
  const type = core.getInput("type");
  const whoToGreet = core.getInput("who-to-greet");
  const message = \`Hello \${whoToGreet}!\`;
  core.info(message);
  if (type === "error") {
    core.setFailed("This is an error!");
  }
  core.setOutput("message", message);
}
run().catch((err) => {
  console.error(err);
  core.setFailed(err.message);
});
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 1`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        
declare global {
  export const ACTION_INPUTS: {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
}

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 2`] = `
""use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var core = __toESM(require("@actions/core"), 1);
globalThis.ACTION_INPUTS = {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
async function run() {
  const type = core.getInput("type");
  const whoToGreet = core.getInput("who-to-greet");
  const message = \`Hello \${whoToGreet}!\`;
  core.info(message);
  if (type === "error") {
    core.setFailed("This is an error!");
  }
  core.setOutput("message", message);
}
run().catch((err) => {
  console.error(err);
  core.setFailed(err.message);
});
;
"
`;
